File: articles/regime-shifts-and-early-warning-signals/fortran/regime_shift_simulation.f90
program regime_shift_simulation
  implicit none

  integer, parameter :: steps = 180
  integer :: t
  real :: pressure, start_pressure, end_pressure, x

  start_pressure = -0.75
  end_pressure = 0.85
  x = -0.90

  print *, "time,pressure,state,regime"

  do t = 1, steps
     pressure = start_pressure + (end_pressure - start_pressure) * real(t - 1) / real(steps - 1)
     if (t > 1) then
        x = x + 0.05 * (1.2 * x - x**3 + pressure)
     end if

     if (x >= 0.0) then
        print '(I0,A,F9.5,A,F9.5,A,A)', t, ",", pressure, ",", x, ",", "upper regime"
     else
        print '(I0,A,F9.5,A,F9.5,A,A)', t, ",", pressure, ",", x, ",", "lower regime"
     end if
  end do
end program regime_shift_simulation

File: articles/regime-shifts-and-early-warning-signals/cpp/regime_shift_simulation.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <string>

double update_state(double x, double pressure, double r = 1.2, double dt = 0.05) {
    return x + dt * (r * x - x * x * x + pressure);
}

int main() {
    const int steps = 180;
    const double start = -0.75;
    const double end = 0.85;

    double x = -0.90;

    std::cout << "time,pressure,state,regime\n";

    for (int t = 1; t <= steps; ++t) {
        double pressure = start + (end - start) * (t - 1) / (steps - 1);
        if (t > 1) {
            x = update_state(x, pressure);
        }

        std::string regime = x >= 0.0 ? "upper regime" : "lower regime";
        std::cout << t << "," << pressure << "," << x << "," << regime << "\n";
    }

    return 0;
}
